8367991: Update RegionBackground tests to use ScreenCaptureTestWatcher utility

jayathirthrao opened the pull request:
We are seeing very rare intermittent failures in RegionBackgroundFillUITest & RegionBackgroundImageUITest tests in linux. We already have [JDK-8328217](https://bugs.openjdk.org/browse/JDK-8328217) to fix this issue.

To identify the root cause better, we want to capture the test content when it is failing using ScreenCaptureTestWatcher utility.

Since we wan to capture the test window before test stage is cleared. We are now clearing the test stage and next test setup. Initially tried updating this behaviour in `VisualTestBase` itself, but that causes tests like `test.robot.test3d.PointLightIlluminationTest` as it overrides the test setup functions.

This change is tested to make sure ScreenCaptureTestWatcher takes appropriate screen capture on failure and CI full headful test run is green.

kevinrushforth reviewed (commented):
This approach requires each test to copy logic from the base class into the test class, so it will be harder to maintain and doesn't scale well.

My recommendation is to either:

1. Make the changes to move the hiding of the changes to the base class methods, change all subclasses that currently have a different `@BeforeEach` method _not_named `doSetup` to rename their `@BeforeEach` method to `doSetup()` and call `super.doSetup()` as the first statement of the overridden method
2. Create a protected utility method in VisualTestBase that a subclass can call.

I think 1 might be better, since it keeps all of the logic in the base class, but 2 would be OK as well, since the actual cleaning of the stages would remain in the base class, but it does require overriding the two methods.

andy-goryachev-oracle commented on tests/system/src/test/java/test/robot/javafx/scene/layout/RegionBackgroundImageUITest.java:
```diff
@@ -39,8 +49,37 @@
 
+    @BeforeEach
+    public void doSetup() {
+        runAndWait(() -> {
```
suggestion:

we've got `Util::shutdown()` which `hide()`s all the windows then calls `Platform::exit`.

we could add `Util::hideAllWindows(Runnable)` that can be used in these tests, and remove stages from the `VisualTestBase class`:

```
    /**
     * This synchronous method first hides all the open {@code Window}s in the platform thread,
     * then invokes {@link Platform.exit()}.
     */
    public static void shutdown() {
        hideAllWindows(Platform::exit);
    }

    /// Hides all open `Window`s then invokes `after` (if non-null) in the FX Application Thread.
    public static void hideAllWindows(Runnable after) {
        runAndWait(() -> {
            List.
                copyOf(Window.getWindows()).
                forEach(Window::hide);
            if (after != null) {
                after.run();
            }
        });
    }
```

What do you think?

andy-goryachev-oracle replied:
last time I checked, no system test needs to specify which windows to hide.

kevinrushforth replied:
Maybe not at `@AfterAll` time, but there are tests that close the active window in `@AfterEach` leaving the primary stage shown.

kevinrushforth replied:
Anyway, this is another reason why this should be a follow-up issue: so we can discuss the requirements of a more general utility.

andy-goryachev-oracle commented on tests/system/src/test/java/test/robot/javafx/scene/layout/RegionBackgroundImageUITest.java:
```diff
@@ -39,8 +49,37 @@
+     */
+    @AfterEach
+    public void doTeardown() {
+
```
please remove empty line

jayathirthrao replied:
The function itself is removed.

jayathirthrao commented:
> This approach requires each test to copy logic from the base class into the test class, so it will be harder to maintain and doesn't scale well.
> 
> My recommendation is to either:
> 
>     1. Make the changes to move the hiding of the changes to the base class methods, change all subclasses that currently have a different `@BeforeEach` method _not_named `doSetup` to rename their `@BeforeEach` method to `doSetup()` and call `super.doSetup()` as the first statement of the overridden method
> 
>     2. Create a protected utility method in VisualTestBase that a subclass can call.
> 
> 
> I think 1 might be better, since it keeps all of the logic in the base class, but 2 would be OK as well, since the actual cleaning of the stages would remain in the base class, but it does require overriding the two methods.

I have followed the 1st recommendation and updated to include main change in base class and now all sub-classes override the base class setup functions with same name. I have checked all classes which inherit from VisualTestBase and updated them.

Also PointLightIlluminationTest maintains first created stage and uses it for all consecutive tests, so added flag in base class to make sure we re-use the stage. CI headful test run with latest update is green.

kevinrushforth commented on tests/system/src/test/java/test/robot/test3d/PointLightIlluminationTest.java:
```diff
@@ -81,6 +82,8 @@ public void setupEach() {
             });
             // Ensure that the scene is really displayed, before the tests begin
             waitFirstFrame();
+            // The same scene can be used for all the illumination tests by this class
```
Minor: "scene" --> "stage" ?

kevinrushforth commented on tests/system/src/test/java/test/robot/testharness/VisualTestBase.java:
```diff
@@ -55,11 +55,13 @@ public abstract class VisualTestBase {
     // Glass Robot instance
     Robot robot;
 
+    protected static boolean clearStage = true;
```
Minor: would `clearStages` (plural) be a better name?

kevinrushforth reviewed (approved):
LGTM. I left a couple minor comments inline and will reapprove if you make any changes.
